src/routing/approved.rs

```diff
@@ -1090,34 +1104,74 @@ impl Approved {
     ) -> Result<Vec<Command>> {
         trace!("Received DKGStart for {}", new_elders_info);
         self.dkg_voter
-            .start_participating(self.node.name(), dkg_key, new_elders_info)
+            .start(&self.node.keypair, dkg_key, new_elders_info)
             .into_commands(&self.node)
     }
 
-    fn handle_dkg_result(
+    fn handle_dkg_message(
         &mut self,
         dkg_key: DkgKey,
-        result: Result<bls::PublicKey, ()>,
+        message: DkgMessage,
         sender: XorName,
     ) -> Result<Vec<Command>> {
+        trace!("handle DKG message {:?} from {}", message, sender);
+
         self.dkg_voter
-            .observe_result(&dkg_key, result, sender)
+            .process_message(&self.node.keypair, dkg_key, message)
             .into_commands(&self.node)
     }
 
-    fn handle_dkg_message(
+    fn handle_dkg_failure_observation(
         &mut self,
         dkg_key: DkgKey,
-        message: DkgMessage,
-        sender: XorName,
+        proof: DkgFailureProof,
     ) -> Result<Vec<Command>> {
-        trace!("handle DKG message {:?} from {}", message, sender);
-
         self.dkg_voter
-            .process_message(&self.node.name(), dkg_key, message)
+            .process_failure(dkg_key, proof)
             .into_commands(&self.node)
     }
 
+    fn handle_dkg_failure_agreement(
+        &self,
+        sender: &XorName,
+        elders_info: EldersInfo,
+        proofs: &DkgFailureProofSet,
+    ) -> Result<Vec<Command>> {
+        let sender = &self
+            .section
+            .members()
+            .get(sender)
+            .ok_or(Error::InvalidSource)?
+            .peer;
+
+        if !proofs.verify(&elders_info) {
+            error!(
+                "Ignore DKG failure agreement with invalid proofs: {}",
+                elders_info
+            );
+            return Ok(vec![]);
+        }
+
+        if !self
+            .section
+            .promote_and_demote_elders(&self.node.name())
+            .contains(&elders_info)
+        {
+            trace!(
+                "Ignore DKG failure agreement for outdated participants: {}",
+                elders_info
+            );
+            return Ok(vec![]);
+        }
+
+        trace!(
+            "Received DKG failure agreement - restarting: {}",
+            elders_info
+        );
+
+        self.send_dkg_start_to(elders_info, slice::from_ref(sender))
```

Comment:
shall the start be sent to ALL elders_info ?
otherwise the chance of `out-of-snyc` could got increased?

also, the log above says `restarting elders_info` , however actually we just restart one of it.

Reply:
It should be enough to send the DKGStart only to those nodes that observed the failure, because eventually all of the participants should observe it and will send the failure agreement too. Sending to all of them would open us up for spam attack, because we would have a situation where a single message triggers multiple messages.